cmd/k8s-operator/tsrecorder_specs.go

```diff
@@ -192,21 +162,117 @@ func tsrStateSecret(tsr *tsapi.TSRecorder, namespace string) *corev1.Secret {
 		ObjectMeta: metav1.ObjectMeta{
 			Name:            fmt.Sprintf("%s-0", tsr.Name),
 			Namespace:       namespace,
-			Labels:          labels("tsrecorder", tsr.Name),
+			Labels:          labels("tsrecorder", tsr.Name, nil),
 			OwnerReferences: tsrOwnerReference(tsr),
 		},
 	}
 }
 
-func labels(app, instance string) map[string]string {
-	// ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/
-	return map[string]string{
-		"app.kubernetes.io/name":       app,
-		"app.kubernetes.io/instance":   instance,
-		"app.kubernetes.io/managed-by": "tailscale-operator",
+func env(tsr *tsapi.TSRecorder) []corev1.EnvVar {
+	envs := []corev1.EnvVar{
+		{
+			Name: "TS_AUTHKEY",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: tsr.Name,
+					},
+					Key: "authkey",
+				},
+			},
+		},
+		{
+			Name: "POD_NAME",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					// Secret is named after the pod.
+					FieldPath: "metadata.name",
+				},
+			},
+		},
+		{
+			Name:  "TS_STATE",
+			Value: "kube:$(POD_NAME)",
+		},
+		{
+			Name:  "TSRECORDER_HOSTNAME",
+			Value: "$(POD_NAME)",
+		},
+	}
+
+	for _, env := range tsr.Spec.StatefulSet.Pod.Container.Env {
+		envs = append(envs, corev1.EnvVar{
+			Name:  string(env.Name),
+			Value: env.Value,
+		})
+	}
+
+	if tsr.Spec.Storage.S3 != nil {
+		envs = append(envs,
+			corev1.EnvVar{
+				Name:  "TSRECORDER_DST",
+				Value: fmt.Sprintf("s3://%s", tsr.Spec.Storage.S3.Endpoint),
+			},
+			corev1.EnvVar{
+				Name:  "TSRECORDER_BUCKET",
+				Value: tsr.Spec.Storage.S3.Bucket,
+			},
+		)
+
+		if tsr.Spec.Storage.S3.CredentialsSecret != "" {
+			envs = append(envs, corev1.EnvVar{
+				Name: "TSRECORDER_DST",
```

Comment:
Oops, I meant to set this in EnvFrom - I just pushed what I had intended in 38774a4af50d7769a795d11650fbd3b495dd91e9. It mounts all the secret data so that people can optionally include a session token as well, or I think they could even just use it to configure non-sensitive metadata for workload identity options. So far I've only tested a static access key though - but that flow is now working (feel free to use my AWS sandbox user for testing).

Reply:
Looks great!